tower/src/util/and_then.rs

```diff
@@ -0,0 +1,81 @@
+use std::marker::PhantomData;
+use std::task::{Context, Poll};
+
+use futures_core::TryFuture;
+use futures_util::{future::AndThen as AndThenFut, TryFutureExt};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Service returned by the [`and_then`] combinator.
+///
+/// [`and_then`]: crate::util::ServiceExt::and_then
+#[derive(Clone, Debug)]
+pub struct AndThen<S, F, Fut> {
+    inner: S,
+    f: F,
+    _fut: PhantomData<Fut>,
+}
```

Comment:
I don't think it's necessary for the `AndThen` type to be generic over the response future's type. If we remove `Fut` from the type of `AndThen` and `AndThenLayer`, we don't need to use `PhantomData`, and the type of the `AndThen` service is smaller. This may have an impact on compile times when there are large nested service stacks.

I can open a PR against your branch with a version that removes the type parameter, if you like --- I made the change myself before commenting to make sure it does, in fact, compile.

Reply:
👍 that'd be awesome. It's only there because I assumed it'd be required from its presence in the futures AndThen. Let's get rid of it.

Reply:
done! https://github.com/hlb8122/tower/pull/1